### ISSUE TITLE
Cannot reliably close WebSocket / coro-websocket read message is never nil

### ISSUE THREAD

**Issue by SinisterRectus** [luvit/lit#208]
In coro-websocket 1.0.0-1, I wrote an empty message to close the socket:

```lua
write()
```
This causes an opcode of 8 to be sent:
```lua
  local function write(message)
    if message then
      ...
    else -- send close frame if message is nil/false
      if not closeSent then
        return write({
          opcode = 8,
          payload = ""
        })
      end
    end
    return rawWrite(message)
  end
```
After this, `nil` was returned to signal an end of data:
```lua
  local function read()
    while true do
      local message = rawRead()
      if not message then
        return cleanup() -- this returns nil to break the loop
      end
...
```
And the read loop breaks:
```lua
for message in read do
  -- handle messages
end
-- handle socket close after loop breaks
```
Similarly, if the server closed the connection, `nil` is returned and the loop breaks.

----

While working with coro-websocket 3.0.0, I noticed that `nil` is not consistently returned.

If the server closes the connection, I get a libuv error and the process hangs:
```
../coro-channel.lua:119: ../coro-channel.lua:103: ENOTCONN: socket is not connected
```

If `res.socket` is closed or if `write()` is called, the process appears to end (crash?) after one read loop or less; code after the loop is never executed.

PR #207 would help with a server close, but I'm not sure what to do if the client wants to close the connection.

**Comment by SinisterRectus** [comment]
I think I've narrowed down the issue to somewhere in coro-wrapper by using the following test:

```lua
coroutine.wrap(function()

	local ws = require('coro-websocket')

	local options = ws.parseUrl("wss://gateway.discord.gg")
	options.pathname = '/?v=7&encoding=json'

	local res, read, write = ws.connect(options)

	for msg in read do
		write()
	end

end)()
```

Looking at `decoder` at line 71:

The final message in `chunk` is `nil`, which causes `buffer` to become an empty string. When this is decoded, both `item` and `newIndex` come back as `nil`. Because they are both `nil`, the `return item` line is never reached:

```lua
if item or newIndex then
  ...
  return item
end
```
 I can think of a few different ways to solve this:

- Move `return item` outside of its current block.
- Add a special case for `buffer == ''`
- Have the decoder to return `nil, 0` instead of nothing.

The first is probably the easiest, but I don't know if it will affect anything else.

**Comment by SinisterRectus** [comment]
I also noticed that, according to the code, both `buffer` and `chuck` can be `nil`. I don't know if this is ever possible in practice, but if it does happen, `decode` will attempt to decode a nil buffer. It looks like the HTTP decoder will return on an empty buffer, but the WebSocket decoder will error. It may be necessary to extend the buffer check:

```lua
if buffer == '' or not buffer then
```

**Comment by creationix** [comment]
It looks like your head is more in this than mine.  I'm super busy with another project for the next few weeks.  Go ahead and PR the one you prefer.  I'll try it against lit's unit tests and see if it breaks anything in there (they use coro websockets heavily).

**Comment by SinisterRectus** [comment]
Addressed with 84fe7b9
